File: packages/shared-go/obs/log.go
package obs

import (
	"context"
	"io"
	"log/slog"
	"os"
	"strings"

	"go.opentelemetry.io/otel/trace"
)

// Setup builds the process-wide *slog.Logger for serviceName.
//
//   - Handler: JSON by default (containers); LOG_FORMAT=text switches to a human-readable handler for
//     local dev.
//   - Level: LOG_LEVEL (debug/info/warn/error, case-insensitive), default info. An unrecognized value
//     falls back to info rather than erroring — a typo'd env var must not crash the service.
//   - Every line carries LogKeyService=serviceName.
//   - Every line logged through a context that carries a span automatically carries LogKeyTraceID/
//     LogKeySpanID (see Handler) — call sites log with *Context methods (InfoContext, ErrorContext,
//     ...) and a context; they must never thread trace/span ids by hand.
func Setup(serviceName string) *slog.Logger {
	return SetupTo(os.Stdout, serviceName)
}

// SetupTo is Setup with an explicit destination. It exists so the "every line carries LogKeyService"
// promise is testable: with Setup writing only to os.Stdout, a mutation that dropped the service attribute
// entirely left the whole unit suite green (found by W0's review). A seam that makes a documented
// guarantee assertable is worth more than one fewer exported function.
func SetupTo(w io.Writer, serviceName string) *slog.Logger {
	opts := &slog.HandlerOptions{Level: parseLevel(os.Getenv("LOG_LEVEL"))}

	var base slog.Handler
	if strings.EqualFold(os.Getenv("LOG_FORMAT"), "text") {
		base = slog.NewTextHandler(w, opts)
	} else {
		base = slog.NewJSONHandler(w, opts)
	}

	handler := NewHandler(base).WithAttrs([]slog.Attr{slog.String(LogKeyService, serviceName)})
	return slog.New(handler)
}

func parseLevel(raw string) slog.Level {
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case "debug":
		return slog.LevelDebug
	case "warn", "warning":
		return slog.LevelWarn
	case "error":
		return slog.LevelError
	default:
		return slog.LevelInfo
	}
}

// Handler wraps an slog.Handler and injects LogKeyTraceID/LogKeySpanID into every record whose context
// carries a span with a valid span context (trace.SpanContextFromContext(ctx).IsValid()) — this covers
// both a locally-started recording span and a span context extracted from an inbound traceparent, which
// is deliberately broader than "only a locally recording span": the trace id is worth logging for
// correlation either way, sampling decision notwithstanding. A context with no span in it (the common
// case for anything logged outside a request/message handler) is passed through unchanged — this is not
// an error, it is simply "no trace in scope right now," and call sites never need to know the
// difference: they just log with a context.
type Handler struct {
	next slog.Handler
}

// NewHandler wraps next. Setup is the common entry point; NewHandler is exported for callers assembling
// a custom handler chain (e.g. a test that wants to inspect records without going through Setup's
// env-driven format/level selection).
func NewHandler(next slog.Handler) *Handler {
	return &Handler{next: next}
}

func (h *Handler) Enabled(ctx context.Context, level slog.Level) bool {
	return h.next.Enabled(ctx, level)
}

func (h *Handler) Handle(ctx context.Context, record slog.Record) error {
	if sc := trace.SpanContextFromContext(ctx); sc.IsValid() {
		record.AddAttrs(
			slog.String(LogKeyTraceID, sc.TraceID().String()),
			slog.String(LogKeySpanID, sc.SpanID().String()),
		)
	}
	return h.next.Handle(ctx, record)
}

func (h *Handler) WithAttrs(attrs []slog.Attr) slog.Handler {
	return &Handler{next: h.next.WithAttrs(attrs)}
}

func (h *Handler) WithGroup(name string) slog.Handler {
	return &Handler{next: h.next.WithGroup(name)}
}

File: apps/dashboard-web/src/lib/server/observability/log.ts
// Structured server-side logger. Every line is one JSON object on stdout carrying `service` and,
// whenever the call happens inside a request wrapped by hooks.server.ts's `runWithTraceContext`,
// `trace_id`/`span_id` — mirroring packages/shared-go/obs's slog.Handler (see log.go), which injects
// the same two keys onto any record logged through a context carrying a span. Call sites here never
// thread the trace id by hand, same guarantee as the Go side.
//
// Server-only (see trace.ts for why). Route handlers, $lib/server/*, and $lib/db/* should import this
// instead of $lib/logger.ts.

import { LOG_KEY_EVENT, LOG_KEY_SERVICE, LOG_KEY_SPAN_ID, LOG_KEY_TRACE_ID, SERVICE_NAME } from './constants';
import { getTraceContext } from './trace';

export type LogFields = Record<string, unknown>;

function serializeError(err: unknown): { message: string; stack?: string } {
	if (err instanceof Error) {
		return { message: err.message, stack: err.stack };
	}
	return { message: String(err) };
}

function emit(level: 'info' | 'warn' | 'error', event: string, fields?: LogFields): void {
	const ctx = getTraceContext();
	const line: Record<string, unknown> = {
		time: new Date().toISOString(),
		level,
		[LOG_KEY_SERVICE]: SERVICE_NAME,
		[LOG_KEY_EVENT]: event,
		...(ctx ? { [LOG_KEY_TRACE_ID]: ctx.traceId, [LOG_KEY_SPAN_ID]: ctx.spanId } : {}),
		...fields,
	};
	// eslint-disable-next-line no-console -- this IS the structured-logging sink (stdout, one JSON line).
	console.log(JSON.stringify(line));
}

export const log = {
	info(event: string, fields?: LogFields): void {
		emit('info', event, fields);
	},
	warn(event: string, fields?: LogFields): void {
		emit('warn', event, fields);
	},
	/** `error` field, if present, is passed through `serializeError` (Error -> {message, stack}, else
	 * String(err)) rather than logged raw — an Error object stringifies to `{}` under JSON.stringify. */
	error(event: string, fields?: LogFields & { error?: unknown }): void {
		const { error: err, ...rest } = fields ?? {};
		emit('error', event, err !== undefined ? { ...rest, error: serializeError(err) } : rest);
	},
};
